lib/SPIRV/SPIRVLowerMemmove.cpp

```diff
@@ -0,0 +1,118 @@
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of Advanced Micro Devices, Inc., nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering llvm.memmove into several llvm.memcpys.
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spvmemmove"
+
+#include "SPIRVInternal.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Pass.h"
+#include "llvm/PassSupport.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace SPIRV;
+
+namespace SPIRV {
+cl::opt<bool> SPIRVLowerMemmoveValidate("spvmemmove-validate",
+    cl::desc("Validate module after lowering llvm.memmove instructions into llvm.memcpy"));
+
+class SPIRVLowerMemmove: public ModulePass,
+  public InstVisitor<SPIRVLowerMemmove> {
+public:
+  SPIRVLowerMemmove():ModulePass(ID), Context(nullptr) {
+    initializeSPIRVLowerMemmovePass(*PassRegistry::getPassRegistry());
+  }
+  virtual void visitMemMoveInst(MemMoveInst &I) {
+    IRBuilder<> Builder(I.getParent());
+    Builder.SetInsertPoint(&I);
+    auto Dest = I.getRawDest();
+    auto Src = I.getRawSource();
+    auto SrcTy = Src->getType();
+    ConstantInt *Length = nullptr;
+    if(isa<ConstantInt>(I.getLength())) //the length (in bytes) could be non-constant
+        Length = ConstantInt::get(Type::getInt64Ty(*Context), cast<ConstantInt>(I.getLength())->getZExtValue());
+    if(isa<BitCastInst>(Src)) //the source could be bit-cast from another type, need the original type for the allocation
+        SrcTy = cast<BitCastInst>(Src)->getOperand(0)->getType();
+    auto Align = I.getAlignment();
+    auto Volatile = I.isVolatile();
+
+    auto Alloca = Builder.CreateAlloca(SrcTy->getPointerElementType(), SrcTy->isArrayTy() ? Builder.getInt32(SrcTy->getArrayNumElements()) : nullptr);

```

Comment:
this line may exceed 80 columns.
Please follow http://llvm.org/docs/CodingStandards.html